src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThreadGroup.java

```diff
@@ -200,8 +204,29 @@ private void assertVmThread() {
         assert current()==this;
     }
 
+    /**
+     * Returns the thread with the specified id.
+     *
+     * Normally called from the CPS VM thread, but may be called from other threads via {@link CpsStepContext#doGet}.
+     *
+     * @return
+     *      null if the thread has finished executing.
+     */
     public CpsThread getThread(int id) {
-        return threads.get(id);
+        CpsThread thread = threads.get(id);
+        if (thread == null && LOGGER.isLoggable(Level.FINE)) {
+            LOGGER.log(Level.FINE, "no thread " + id + " among " + threads.keySet(), new IllegalStateException());
+        }
+        return thread;
+    }
+
+    /**
+     * Returns an unmodifiable view of all threads in the thread group.
+     */
+    @CpsVmThreadOnly
+    public Collection<CpsThread> getThreads() {
```

Comment:
I introduced this method to be able to make `threads` private, and to make it easy to assert that all code iterating over threads is doing so in the CPS VM thread and so doesn't need any additional synchronization.

Reply:
I'd go with `Iterable` by default (i.e. unless some specialization in `Collection` is needed -- this usually being `Collection::size`).

Reply:
I went with `Iterable` in cb6b2b2 since all existing uses are just for-each loops.